METAPROGRAMMING.md

```diff
@@ -1,86 +1,329 @@
+    assert(kind == 'class');
     return {
       kind,
       elements,
       // This callback is called once the class is otherwise fully defined
       finisher(klass) {
         window.customElements.define(tagName, klass);
       }
-    }
-  }
+    };
+  };
 }
 
 // Replace a method with a field with a bound version of the method
 function bound(elementDescriptor) {
-  let {kind, key, placement, descriptor} = elementDescriptor;
-  assert(kind === "method");
-  if (placement == "prototype") placement = "own";
-  const {value} = descriptor;
-  function initializer() { return value.bind(this); }
+  let { kind, key, placement, descriptor } = elementDescriptor;
+  assert(kind === 'method');
+  if (placement == 'prototype') placement = 'own';
+  const { value } = descriptor;
+  function initializer() {
+    return value.bind(this);
+  }
   delete descriptor.value;
-  return { kind: "field", key, placement, descriptor, initializer };
+  return { kind: 'field', key, placement, descriptor, initializer };
 }
 
 // Whenever a read or write is done to a field, call the render()
 // method afterwards. Implement this by replacing the field with
 // a getter/setter pair.
-function observed({kind, key, placement, descriptor, initializer}) {
-  assert(kind == "field");
-  assert(placement == "own");
+function observed({ kind, key, placement, descriptor, initializer }) {
+  assert(kind == 'field');
+  assert(placement == 'own');
   // Create a new anonymous private name as a key for a class element
   let storage = decorators.PrivateName();
-  let underlyingDescriptor = { enumerable: false, configurable: false, writable: true };
-  let underlying = { kind, key: storage, placement, underlyingDescriptor, initializer };
+  let underlyingDescriptor = {
+    enumerable: false,
+    configurable: false,
+    writable: true
+  };
+  let underlying = {
+    kind,
+    key: storage,
+    placement,
+    underlyingDescriptor,
+    initializer
+  };
   return {
-    kind: "method",
+    kind: 'method',
     key,
     placement,
     descriptor: {
-      get() { storage.get(this); },
+      get() {
+        storage.get(this);
+      },
       set(value) {
         storage.set(this, value);
         // Assume the @bound decorator was used on render
         window.requestAnimationFrame(this.render);
       },
       enumerable: descriptor.enumerable,
-      configurable: descriptor.configurable,
+      configurable: descriptor.configurable
     },
     extras: [underlying]
   };
-  
 }
 ```
+
+## Class element finisher
+
+Here is an example that uses the `finisher` property for a class element:
+
+```js
+class User {
+  @readonly id;
+  email;
+
+  constructor(id = null, email = null) {
+    this.id = id;
+    this.email = email;
+  }
+}
+
+function readonly(elementDecriptor) {
+  // we don't want to set `writable` to false until after the property has been initialized
+  elementDecriptor.finisher = () => {
+    elementDecriptor.descriptor.writable = false;
```

Comment:
This won't actually work--nothing will be paying attention to the element descriptor at this time. Anyway, the finisher runs once for the whole class, not per instance.

Reply:
I could think of a different example of a class element finisher. But wouldn't finishers be more useful if they had access to `this` for class element decorators? (This is only applicable if `placement` is ultimately set to `own` of course.) Otherwise I'm not sure how I would even implement my read-only decorator. For `babel-plugin-transform-decorators-legacy`, I implemented the `@readonly` decorator to store the names of all the read-only properties in the prototype, and designed it to work in conjunction with a particular class that knows to look for those properties and set their initial values from an attributes object passed to the constructor. (That could be generalized to all classes by additionally requiring the use of a class-level decorator.) But it seems that in the current proposal, it might be even harder to implement, because you can't even directly access the prototype from a decorator function. Don't get me wrong, I think there are many great improvements in the current proposal, but isn't there some way this use case could be better supported?

Reply:
For now I just removed the `@readonly` example from the PR.

Reply:
One other thing: if a finisher is only run once per class, then isn't this misleading: "A callback that is called with the constructor, once it's created." I copied that from the existing document.

Reply:
Probably there are ways that the existing document could be clarified; it seems like "it" is a bit ambiguous there. The finisher is called once for the whole class. There is an idea to have "instance finishers", but it is not as well-developed as the rest of this proposal and we're leaving for v2.

Reply:
How about: "A callback that is called at the end of class creation (before class-level finishers, if they exist)"